File: 25-reverse-nodes-in-k-group/25-reverse-nodes-in-k-group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    boolean isSufficientNode(ListNode head,int k){
        ListNode curr = head;
        while(k > 0 && curr != null){
            curr = curr.next;
            k--;
        }
        return (k == 0);
    }
    ListNode reverseGroup(ListNode head,int k){
        if(!isSufficientNode(head,k)){
            return head;
        }
        
        ListNode curr = head;
        ListNode start = head;
        ListNode prev = null;
        for(int i=1;i<=k;i++){
            ListNode nxt = curr.next;
            curr.next = prev;
            prev = curr;
            curr = nxt;
        }
        
        start.next = reverseGroup(curr,k);
        return prev;
    }
    public ListNode reverseKGroup(ListNode head, int k) {
        
        return reverseGroup(head,k);
    }
}

File: 25-reverse-nodes-in-k-group/25-reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    
    int len(ListNode *head) {
        if(head == NULL) return 0;
        return len(head->next) + 1;
    }
    ListNode* reverseKGroup(ListNode* head, int k) {
        
        if(head == NULL) return NULL;
        if(len(head) < k) return head;
        ListNode *curr = head;
        ListNode *prev = NULL;
        int count = 0;
        while(count < k && curr != NULL) {
            ListNode *nxt = curr->next;
            
            curr->next = prev;
            prev = curr;
            curr = nxt;
            count++;
        }
        
        head->next  = reverseKGroup(curr,k);
        return prev;
    }
};
